From 8babf0e548b3a391ccaa365ad65260b961afe814 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Fri, 31 May 2019 07:35:19 +0200
Subject: [PATCH 1/1] efi_loader: check time in SetTime()

The UEFI spec prescribes that we check that the timestamp passed to
SetTime() is checked for validity.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 include/efi_loader.h         | 10 ++++++++++
 lib/efi_loader/efi_runtime.c | 22 +++++++++++++++++++++-
 2 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/include/efi_loader.h b/include/efi_loader.h
index 77b2f60bdc..cb30000d1f 100644
--- a/include/efi_loader.h
+++ b/include/efi_loader.h
@@ -590,6 +590,16 @@ void __efi_runtime EFIAPI efi_reset_system(
 /* Architecture specific initialization of the EFI subsystem */
 efi_status_t efi_reset_system_init(void);
 
+/**
+ * efi_validate_time() - checks if timestamp is valid
+ *
+ * This helper function can be used to implement efi_set_time().
+ *
+ * @time:	timestamp to validate
+ * Returns:	0 if timestamp is valid, 1 otherwise
+ */
+int __efi_runtime efi_validate_time(struct efi_time *time);
+
 efi_status_t __efi_runtime EFIAPI efi_get_time(
 			struct efi_time *time,
 			struct efi_time_cap *capabilities);
diff --git a/lib/efi_loader/efi_runtime.c b/lib/efi_loader/efi_runtime.c
index 058b40a887..a3f7bc4d76 100644
--- a/lib/efi_loader/efi_runtime.c
+++ b/lib/efi_loader/efi_runtime.c
@@ -214,6 +214,26 @@ out:
 #endif
 }
 
+/**
+ * efi_validate_time() - checks if timestamp is valid
+ *
+ * @time:	timestamp to validate
+ * Returns:	0 if timestamp is valid, 1 otherwise
+ */
+int __efi_runtime efi_validate_time(struct efi_time *time)
+{
+	return (!time ||
+		time->year < 1900 || time->year > 9999 ||
+		time->month < 1 || time->month > 12 ||
+		time->day < 1 || time->day > 31 ||
+		time->hour > 23 || time->minute > 59 || time->second > 59 ||
+		time->nanosecond > 999999999 ||
+		time->daylight &
+		~(EFI_TIME_IN_DAYLIGHT | EFI_TIME_ADJUST_DAYLIGHT) ||
+		((time->timezone < -1440 || time->timezone > 1440) &&
+		time->timezone != EFI_UNSPECIFIED_TIMEZONE));
+}
+
 /**
  * efi_set_time_boottime() - set current time
  *
@@ -235,7 +255,7 @@ static efi_status_t EFIAPI efi_set_time_boottime(struct efi_time *time)
 
 	EFI_ENTRY("%p", time);
 
-	if (!time) {
+	if (efi_validate_time(time)) {
 		ret = EFI_INVALID_PARAMETER;
 		goto out;
 	}
-- 
2.20.1

